packages/evm/src/precompiles/bls12_381/util.ts

```diff
@@ -0,0 +1,86 @@
+import { equalsBytes, short } from '@ethereumjs/util'
+
+import { BLS_GAS_DISCOUNT_PAIRS } from './constants.js'
+
+import type { PrecompileInput } from '../types'
+
+const ZERO_BYTES_16 = new Uint8Array(16)
+
+export const gasCheck = (opts: PrecompileInput, gasUsed: bigint, pName: string) => {
+  if (opts._debug !== undefined) {
+    opts._debug(
+      `Run ${pName} precompile data=${short(opts.data)} length=${opts.data.length} gasLimit=${
+        opts.gasLimit
+      } gasUsed=${gasUsed}`
+    )
+  }
+  if (opts.gasLimit < gasUsed) {
+    if (opts._debug !== undefined) {
+      opts._debug(`${pName} failed: OOG`)
+    }
+    return false
+  }
+  return true
+}
+
+export const msmGasUsed = (numPairs: number, gasUsedPerPair: bigint) => {
+  const gasDiscountMax = BLS_GAS_DISCOUNT_PAIRS[BLS_GAS_DISCOUNT_PAIRS.length - 1][1]
+  let gasDiscountMultiplier
+
+  if (numPairs <= BLS_GAS_DISCOUNT_PAIRS.length) {
+    if (numPairs === 0) {
+      gasDiscountMultiplier = 0 // this implicitly sets gasUsed to 0 as per the EIP.
+    } else {
+      gasDiscountMultiplier = BLS_GAS_DISCOUNT_PAIRS[numPairs - 1][1]
+    }
+  } else {
+    gasDiscountMultiplier = gasDiscountMax
+  }
+  // (numPairs * multiplication_cost * discount) / multiplier
+  return (BigInt(numPairs) * gasUsedPerPair * BigInt(gasDiscountMultiplier)) / BigInt(1000)
+}
+
+export const equalityLengthCheck = (opts: PrecompileInput, length: number, pName: string) => {
+  if (opts.data.length !== length) {
+    if (opts._debug !== undefined) {
+      opts._debug(
+        `${pName} failed: Invalid input length length=${opts.data.length} (expected: ${length})`
+      )
+    }
+    return false
+  }
+  return true
+}
+
+export const moduloLengthCheck = (opts: PrecompileInput, length: number, pName: string) => {
+  if (opts.data.length % length !== 0) {
+    if (opts._debug !== undefined) {
+      opts._debug(
+        `${pName} failed: Invalid input length length=${opts.data.length} (expected: ${length}*k bytes)`
+      )
+    }
+    return false
+  }
+  return true
+}
+
+export const zeroByteCheck = (
+  opts: PrecompileInput,
+  zeroByteRanges: number[][],
```

Comment:
Note: this will thus slice the array **exclusive** of `zeroByteRanges[x][1]`. So for instance the array `[[0,1]]` will only check the 0th byte, not the byte at position 1. Is this correct? (And can docs be added for this method?)

Reply:
Yes, I think this should be all correct, actually this only copies your former code into a new function 😂, see this commit https://github.com/ethereumjs/ethereumjs-monorepo/pull/3471/commits/62f15f2c9da06ef81e2ff92fea0d72d344c9f076.

So this is a BLS specific zero byte check, applying this part of the [EIP](https://eips.ethereum.org/EIPS/eip-2537#field-elements-encoding):

![grafik](https://github.com/ethereumjs/ethereumjs-monorepo/assets/931137/8953b8aa-a335-40b5-9e63-bf75e16cf9bf)

And specificall this "Due to the size of p, the top 16 bytes are always zeroes." part, see that the input subarrays always have a length of 16 bytes.

I'll give this and the other methods some documentation.